include/kernels/madwf_transfer.cuh

```diff
@@ -0,0 +1,150 @@
+
+      static constexpr bool dagger = dagger_;
+
+      F out;      // output vector field
+      const F in; // input vector field
+
+      const int Ls_out; // length of 5th dimension of the out field
+      const int Ls_in;  // length of 5th dimension of the in field
+
+      const int volume_4d_cb;
+
+      const matrix_type *wm_p;
+
+      const int nParity;
+
+      Transfer5DArg(ColorSpinorField &out, const ColorSpinorField &in, const matrix_type *wm_p) :
+        kernel_param(dim3(out.VolumeCB() / out.X(4), out.X(4), out.SiteSubset())),
+        out(out),
+        in(in),
+        Ls_out(out.X(4)),
+        Ls_in(in.X(4)),
+        volume_4d_cb(in.VolumeCB() / in.X(4)),
+        wm_p(wm_p),
+        nParity(in.SiteSubset())
+      {
+
+        if (volume_4d_cb != static_cast<int>(out.VolumeCB() / Ls_out)) {
+          errorQuda("Input and Output fields should have the same 4d volume: %d neq %d.\n", volume_4d_cb,
+                    static_cast<int>(out.VolumeCB() / Ls_out));
+        }
+
+        if (in.Nspin() != 4) errorQuda("nSpin = %d not support", in.Nspin());
+        if (in.Ncolor() != 3) errorQuda("nColor = %d not support", in.Ncolor());
+        if (out.Nspin() != 4) errorQuda("nSpin = %d not support", out.Nspin());
+        if (out.Ncolor() != 3) errorQuda("nColor = %d not support", out.Ncolor());
+
+        if (!in.isNative() || !out.isNative())
+          errorQuda("Unsupported field order out=%d in=%d\n", out.FieldOrder(), in.FieldOrder());
+      }
+    };
+
+    template <class Arg> struct Transfer5D {
+
+      const Arg &arg;
+      constexpr Transfer5D(const Arg &arg) : arg(arg) { }
+      static constexpr const char *filename() { return KERNEL_FILE; }
+
+      /**
+        @brief Apply the Ls_out-by-Ls_in matrix to the input vector
+        @param[in] parity Parity we are on
+        @param[in] x_b Checkerboarded 4-d space-time index
+        @param[in] s The output Ls dimension coordinate
+       */
+      __device__ __host__ inline void operator()(int x_cb, int s, int parity)
+      {
+        constexpr bool dagger = Arg::dagger;
+
+        using real = typename Arg::real;
+        using Vector = typename Arg::Vector;
+        using matrix_type = typename Arg::matrix_type;
+
+        const int Ls_in = arg.Ls_in;
+        const int Ls_out = arg.Ls_out;
+        const int volume_4d_cb = arg.volume_4d_cb;
+        const matrix_type *wm_p = arg.wm_p;
+
+        int thread_idx = target::thread_idx().y * target::block_dim().x + target::thread_idx().x;
+        SharedMemoryCache<real> cache;
+        while (thread_idx < static_cast<int>(Ls_out * Ls_in * sizeof(matrix_type) / sizeof(real))) {
+          cache.data()[thread_idx] = reinterpret_cast<const real *>(wm_p)[thread_idx];
+          thread_idx += target::block_dim().y * target::block_dim().x;
+        }
+        cache.sync();
+
+        if (x_cb >= volume_4d_cb) return;
+        if (s >= Ls_out) return;
+        if (parity >= arg.nParity) return;
```

Comment:
If these are the thread dimensions that are passed to the kernel, e.g., `kernel_param.threads` then this bound checking is unnecessary here and will be done already in the generic kernel